File: guests/ocaml/sizepolicy.ml
(* The sizepolicy scene, OCaml port — guests/rust/sizepolicy.rs,
   tools/scenes/sizepolicy.steps. *)

(* No [open Kaya_wire]: this scene names no wire value constructor. *)
open Kaya_app

(* The declared box of the two CONSTANT-mode canvases: the one number
   `scale` and `fixed` disagree about. *)
let box : viewbox = (300.0, 120.0)

(* A rectangle at [l..r] and [t..b] as FRACTIONS of the box. *)
let panel d ((w, h) : viewbox) l t r b paint =
  move_to d (l *. w) (t *. h);
  line_to d (r *. w) (t *. h);
  line_to d (r *. w) (b *. h);
  line_to d (l *. w) (b *. h);
  close d;
  fill d ~paint ()

(* The figure the three drawing canvases share. The centre probe point is
   opaque, which is what [expect_ink] rests on. *)
let figure d box =
  panel d box 0.05 0.0 0.95 1.0 Ground;
  panel d box 0.25 0.0 0.75 1.0 Series_fill

(* The bar whose RIGHT EDGE is the frame number; the scene pins exact frames. *)
let bar d box frame =
  panel d box 0.25 0.0 (0.35 +. (0.10 *. float_of_int frame)) 1.0 Axis

(* Seconds back to the frame the harness drove, off the time the guest was
   HANDED and never one of its own (§15.4). *)
let frame_of time = int_of_float (Float.max 0.0 (Float.round (time *. 60.0)))

let () =
  let app = Kaya_app.create () in

  build app (fun () ->
     window ~title:"sizepolicy" ~width:480.0 ~height:420.0 ();

     let root =
       column
         [
           (* SCALE (the default) *)
           canvas ~grow:1.0 ~a11y_id:"fit" ~a11y_label:"Scaled panel"
             ~viewbox:box ~draw:(fun d -> figure d box);
           (* FIXED *)
           canvas ~grow:1.0 ~a11y_id:"mark" ~a11y_label:"Fixed mark"
             ~viewbox:box ~draw:(fun d -> figure d box) ~fixed:true;
           (* REDRAW *)
           canvas ~grow:1.0 ~a11y_id:"live" ~a11y_label:"Redrawn panel"
             ~viewbox:box ~on_draw:(fun d size -> figure d size);
           (* TICK: under the harness the clock is the core's own step. *)
           canvas ~grow:1.0 ~a11y_id:"clock" ~a11y_label:"Animated bar"
             ~viewbox:box
             ~on_tick:(fun d size time -> bar d size (frame_of time));
         ]
         ()
     in
     mount root);

  exit (run app)

File: guests/swift/sizepolicy.swift
// The sizepolicy scene, Swift port — guests/rust/sizepolicy.rs,
// tools/scenes/sizepolicy.steps.

import Foundation

let app = KayaApp()

/// The declared box of the two CONSTANT-mode canvases: the one number `scale`
/// and `fixed` disagree about.
let box = KayaViewbox(300.0, 120.0)

/// A rectangle at `l..r` and `t..b` as FRACTIONS of the box.
func panel(
    _ d: KayaDraw, _ box: KayaViewbox,
    _ l: Double, _ t: Double, _ r: Double, _ b: Double, _ paint: KayaPaint
) {
    let (w, h) = (box.w, box.h)
    d.moveTo(l * w, t * h)
        .lineTo(r * w, t * h)
        .lineTo(r * w, b * h)
        .lineTo(l * w, b * h)
        .close()
        .fill(paint, rule: .nonzero)
}

/// The figure the three drawing canvases share. The centre probe point is
/// opaque, which `expect_ink` rests on, and EVERY VERTICAL EDGE IS 0 OR 1:
/// `expect_drawing` rounds ink bounds to hundredths, so an inset horizontal
/// edge would round away (the Rust guest derives it).
func figure(_ d: KayaDraw, _ box: KayaViewbox) {
    panel(d, box, 0.05, 0.0, 0.95, 1.0, .ground)
    panel(d, box, 0.25, 0.0, 0.75, 1.0, .seriesFill)
}

/// The bar whose RIGHT EDGE is the frame number; the scene asserts exact
/// frames.
func bar(_ d: KayaDraw, _ box: KayaViewbox, _ frame: UInt32) {
    let right = 0.35 + 0.10 * Double(frame)
    panel(d, box, 0.25, 0.0, right, 1.0, .axis)
}

/// Seconds back to the frame the harness drove, off the time the guest was
/// HANDED and never a clock of its own.
func frameOf(_ time: Double) -> UInt32 {
    UInt32(max((time * 60.0).rounded(), 0.0))
}

app.build { tx in
    tx.window(title: "sizepolicy", width: 480, height: 420)
    let root = tx.column {
        // SCALE (the default)
        let fit = tx.canvas(box, grow: 1.0)
        tx.setA11yId(fit, "fit")
        tx.setA11yLabel(fit, "Scaled panel")
        tx.draw(fit) { d in figure(d, box) }

        // FIXED
        let mark = tx.canvas(box, grow: 1.0).fixed()
        tx.setA11yId(mark, "mark")
        tx.setA11yLabel(mark, "Fixed mark")
        tx.draw(mark) { d in figure(d, box) }

        // REDRAW
        let live = tx.canvas(box, grow: 1.0)
        tx.setA11yId(live, "live")
        tx.setA11yLabel(live, "Redrawn panel")
        live.onDraw { d, size in figure(d, size) }

        // TICK: under the harness the clock is the core's own step.
        let clock = tx.canvas(box, grow: 1.0)
        tx.setA11yId(clock, "clock")
        tx.setA11yLabel(clock, "Animated bar")
        clock.onTick { d, size, time in bar(d, size, frameOf(time)) }
    }
    tx.mount(root)
}

app.run()
